tests/image/test_push.py

```diff
@@ -29,6 +32,59 @@ def test_cli_with_apikey_login(docker_client_mock, test_mock):
 
 
 @pytest.mark.usefixtures('project_dir')
+def test_cli_with_progress(docker_client_mock, test_mock):
+    docker_client_mock.login.return_value = {"Status": "Login Succeeded"}
+    docker_client_mock.push.return_value = [
+        {"status": "The push refers to a repository [some/image]"},
+        {"status": "Preparing", "progressDetail": {}, "id": "abc"},
+        {"status": "Preparing", "progressDetail": {}, "id": "def"},
+        {"status": "Preparing", "progressDetail": {}, "id": "egh"},
+        {"status": "Waiting", "progressDetail": {}, "id": "abc"},
+        {"status": "Waiting", "progressDetail": {}, "id": "egh"},
+        {"status": "Pushing", "progressDetail": {"current": 512, "total": 24803}, "id": "abc"},
+        {"status": "Layer already exists", "progressDetail": {}, "id": "def"},
+        {"status": "Pushing", "progressDetail": {"current": 57344, "total": 26348}, "id": "egh"},
+        {"status": "Pushed", "progressDetail": {}, "id": "egh"},
+        {"status": "Pushing", "progressDetail": {"current": 24805, "total": 24803}, "id": "abc"},
+        {"status": "Pushed", "progressDetail": {}, "id": "abc"},
+        {"status": "Successfully pushed"}
+    ]
+    runner = CliRunner()
+    result = runner.invoke(cli, ["dev", "--version", "test"])
+    assert result.exit_code == 0
+    assert result.output.startswith(
+        'Login to registry succeeded.\n'
+        'Pushing registry/user/project:test to the registry.\n\r'
+    )
+
+    tqdm_ascii = is_tqdm_in_ascii_mode(runner)
+    expected = _expected_push_progress(tqdm_ascii)
+    matched = re.search(expected, result.output)
+    assert matched
+    # to make sure the test regexp below is correct
+    assert matched.group(0).startswith('Layers:  33%')
+    assert matched.group(0).endswith('pushed successfully.\n')
+
+
+def _expected_push_progress(ascii=False):
+    # the following string is regexp because push tqdm speed estimation
+    # depends on time and always different
+    expected = (
+        'Layers:  33%\|███▎      \| 1/3\r          \r'
+        'Layers:  67%\|██████▋   \| 2/3\r'
+        'Layers: 100%\|██████████\| 3/3\n\r'
+        'abc: 100%\|██████████\| 24.8K/24.8K \[[?.0-9]*[KM]?B/s\]\n\r'
+        'egh: 100%\|██████████\| 26.3K/26.3K \[[?.0-9]*[KM]?B/s\]\n'
+        'The image registry/user/project:test pushed successfully\.\n'
+    )
+    if ascii:
+        to_replace = {'█': '#', '▎': '3', '▋': '6'}
+        for k in to_replace:
+            expected = expected.replace(k, to_replace[k])
```

Comment:
Also I'd move ascii check inside it - so you would wrap output with some test function and it'll do all the replacement in the background if necessary. If we add more tests that will reduce amount of boilerplate code.